app/rust/src/parser/spending_condition.rs

```diff
@@ -46,102 +31,267 @@ impl TransactionPublicKeyEncoding {
+    }
+
+    pub fn fee(&self) -> Result<u64, ParserError> {
+        be_u64::<'a, ParserError>(&self.data[29..])
+            .map(|res| res.1)
+            .map_err(|_| ParserError::parser_unexpected_value)
+    }
+
+    #[inline(never)]
     pub fn nonce_str(&self) -> Result<ArrayVec<[u8; zxformat::MAX_STR_BUFF_LEN]>, ParserError> {
         let mut output = ArrayVec::from([0u8; zxformat::MAX_STR_BUFF_LEN]);
-        let len = if cfg!(test) {
-            zxformat::u64_to_str(&mut output[..zxformat::MAX_STR_BUFF_LEN], self.nonce)? as usize
-        } else {
-            unsafe {
-                fp_uint64_to_str(
-                    output.as_mut_ptr() as _,
-                    zxformat::MAX_STR_BUFF_LEN as u16,
-                    self.nonce,
-                    0,
-                ) as usize
-            }
-        };
+        let nonce = self.nonce()?;
+        let len = zxformat::u64_to_str(&mut output[..zxformat::MAX_STR_BUFF_LEN], nonce)? as usize;
         unsafe {
             output.set_len(len);
         }
         Ok(output)
     }
 
+    #[inline(never)]
     pub fn fee_str(&self) -> Result<ArrayVec<[u8; zxformat::MAX_STR_BUFF_LEN]>, ParserError> {
         let mut output = ArrayVec::from([0u8; zxformat::MAX_STR_BUFF_LEN]);
-        let len = if cfg!(test) {
-            zxformat::fpu64_to_str(output.as_mut(), self.fee_rate, 0)? as usize
-        } else {
-            unsafe {
-                fp_uint64_to_str(
-                    output.as_mut_ptr() as _,
-                    zxformat::MAX_STR_BUFF_LEN as u16,
-                    self.fee_rate,
-                    0,
-                ) as usize
-            }
-        };
+        let fee = self.fee()?;
+        let len = zxformat::u64_to_str(output.as_mut(), fee)? as usize;
         unsafe {
             output.set_len(len);
         }
         Ok(output)
     }
 }
 
+#[repr(C)]
+#[derive(PartialEq, Debug)]
+pub struct SinglesigSpendingCondition<'a>(&'a [u8]);
+// pub key_encoding: TransactionPublicKeyEncoding,
+// pub signature: MessageSignature<'a>,
+
+/// A structure that encodes enough state to authenticate
+/// a transaction's execution against a Stacks address.
+/// public_keys + signatures_required determines the Principal.
+/// nonce is the "check number" for the Principal.
+#[repr(C)]
+#[derive(Debug, PartialEq)]
+pub struct MultisigSpendingCondition<'a>(&'a [u8]);
+
+#[repr(C)]
+#[derive(Debug, PartialEq)]
+pub enum SpendingConditionSignature<'a> {
+    Singlesig(SinglesigSpendingCondition<'a>),
+    Multisig(MultisigSpendingCondition<'a>),
+}
+
+#[repr(C)]
+#[derive(Debug, PartialEq)]
+pub struct TransactionSpendingCondition<'a> {
+    signer: SpendingConditionSigner<'a>,
+    signature: SpendingConditionSignature<'a>,
+}
+
+impl<'a> SinglesigSpendingCondition<'a> {
+    #[inline(never)]
+    pub fn from_bytes(bytes: &'a [u8]) -> nom::IResult<&[u8], Self, ParserError> {
+        let len = SIGNATURE_LEN as usize + 1;
```

Comment:
why we take 1 more byte? Is there some kind of termination?
Maybe a comment would help?